src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/FilterClause.java

```diff
@@ -12,6 +20,112 @@
     example = """
              {"name": "Aaron", "country": "US"}
               """)
-public record FilterClause(LogicalExpression logicalExpression) {}
+public record FilterClause(LogicalExpression logicalExpression) {
+  public void validate(CommandContext commandContext) {
+    DocumentProjector indexingProjector = commandContext.indexingProjector();
+    // If nothing specified, everything indexed
+    if (indexingProjector.equals(DocumentProjector.identityProjector())) {
+      return;
+    }
+    validateLogicalExpression(logicalExpression, indexingProjector);
+  }
 
-// public record FilterClause(List<ComparisonExpression> comparisonExpressions) {}
+  public void validateLogicalExpression(
+      LogicalExpression logicalExpression, DocumentProjector indexingProjector) {
+    for (LogicalExpression subLogicalExpression : logicalExpression.logicalExpressions) {
+      validateLogicalExpression(subLogicalExpression, indexingProjector);
+    }
+    for (ComparisonExpression subComparisonExpression : logicalExpression.comparisonExpressions) {
+      validateComparisonExpression(subComparisonExpression, indexingProjector);
+    }
+  }
+
+  public void validateComparisonExpression(
+      ComparisonExpression comparisonExpression, DocumentProjector indexingProjector) {
+    String path = comparisonExpression.getPath();
+    boolean isPathIndexed = indexingProjector.isPathIncluded(path);
+
+    // If path is "_id" and it's denied, the operator can only be $eq or $in
+    if (path.equals(DocumentConstants.Fields.DOC_ID) && !isPathIndexed) {
+      FilterOperator filterOperator = comparisonExpression.getFilterOperations().get(0).operator();
+      // if operator is not $eq or $in, throw error
+      if (!filterOperator.equals(ValueComparisonOperator.EQ)
+          && !filterOperator.equals(ValueComparisonOperator.IN)) {
+        throw new JsonApiException(
+            ErrorCode.ID_NOT_INDEXED,
+            String.format(
+                "%s: The filter path ('%s') is not indexed, you can only use $eq or $in as the operator",
+                ErrorCode.ID_NOT_INDEXED.getMessage(), DocumentConstants.Fields.DOC_ID));
+      } else {
+        // else, _id can be used, return
+        return;
+      }
+    }
+
+    JsonLiteral<?> operand = comparisonExpression.getFilterOperations().get(0).operand();
+    // If path is an object (like address), validate the incremental path (like address.city)
+    if (operand.type() == JsonType.ARRAY || operand.type() == JsonType.SUB_DOC) {
+      if (operand.value() instanceof HashMap<?, ?> map) {
```

Comment:
Let's not check for `HashMap` but just `Map` -- in case we wanted to at some point use, say, `TreeMap` (I think actual type might even be `LinkedHashMap` if deserialized by Jackson).
And then cast to `Map<?, ?>`.